staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go

```diff
@@ -177,32 +177,34 @@ func writeLatencyToAnnotation(ctx context.Context, ev *auditinternal.Event) {
 	audit.AddAuditAnnotationsMap(ctx, layerLatencies)
 }
 
-func processAuditEvent(ctx context.Context, sink audit.Sink, ev *auditinternal.Event, omitStages []auditinternal.Stage) bool {
+func processAuditEvent(ctx context.Context, sink audit.Sink, omitStages []auditinternal.Stage) bool {
+	ac := audit.AuditContextFrom(ctx)
 	for _, stage := range omitStages {
-		if ev.Stage == stage {
+		if ac.GetEventStage() == stage {
 			return true
 		}
 	}
 
 	switch {
-	case ev.Stage == auditinternal.StageRequestReceived:
-		ev.StageTimestamp = metav1.NewMicroTime(ev.RequestReceivedTimestamp.Time)
-	case ev.Stage == auditinternal.StageResponseComplete:
-		ev.StageTimestamp = metav1.NewMicroTime(time.Now())
-		writeLatencyToAnnotation(ctx, ev)
+	case ac.GetEventStage() == auditinternal.StageRequestReceived:
+		ac.SetEventStageTimestamp(metav1.NewMicroTime(ac.GetEventRequestReceivedTimestamp().Time))
+	case ac.GetEventStage() == auditinternal.StageResponseComplete:
+		ac.SetEventStageTimestamp(metav1.NewMicroTime(time.Now()))
+		writeLatencyToAnnotation(ctx)
 	default:
-		ev.StageTimestamp = metav1.NewMicroTime(time.Now())
+		ac.SetEventStageTimestamp(metav1.NewMicroTime(time.Now()))
 	}
 
 	audit.ObserveEvent(ctx)
-	return sink.ProcessEvents(ev)
+	return ac.Invoke(func(e *auditinternal.Event) bool {
```

Comment:
Rather than calling invoke here and processing the audit events in locked code, I think it might be better to expose a method on the AuditContext that returns a deep-copied audit event, and just set the stage & stage timestamp on that copied event (since those shouldn't be persisted anyway... they just have to do with the stage it's being output at). I think this should eliminate the need for a bunch of setters & getters on the audit context, and also avoids exposing the internal `visitEvent` functionality.

Reply:
TODO(dims): Need to do this!

Reply:
@tallclair made this change, PTAL. 

`Invoke` i can't get rid of unless we change existing test cases which are sprinkled across packages. same with a bunch of the Setters and Getters. not too worried about `visitEvent` itself as it is still internal to just the `audit` package. right?

Reply:
marking as resolved for now